thin-replica-server/include/thin-replica-server/thin_replica_impl.hpp

```diff
@@ -933,72 +933,84 @@ class ThinReplicaImpl {
     LOG_INFO(logger_, "Sync reading event groups from KVB [" << start << ", " << end << "]");
     readAndSendEventGroups<ServerContextT, ServerWriterT, DataT>(logger_, context, stream, start, end, kvb_filter);
 
-    auto last_global_eg_id_read = kvb_filter->getLastGlobalEgIdReadForClient();
-    uint64_t newest_eg_id = 0;
-    // Let's wait until we have at least one live update for the requesting client
     bool is_update_available = false;
+    int num_updates_filtered_out = 0;
+    auto next_global_eg_id_to_read = kvb_filter->getLastGlobalEgIdReadForClient() + 1;
     while (not is_update_available) {
       is_update_available = live_updates->waitForEventGroupUntilNonEmpty(kWaitForUpdateTimeout);
       if (context->IsCancelled()) {
         throw StreamCancelled("StreamCancelled while waiting for the first live update");
       }
-      if (is_update_available) {
-        newest_eg_id = kvb_filter->newestExternalEventGroupId();
-        // There are two scenarios in which updates will be dropped from the live-update-queue
-        // 1. If the oldest global-eg-id on the live-update-queue is less than last_global_eg_id_read + 1
-        // 2. If the update available on the live-update-queue is for another client (i.e., newest_eg_id in storage is
-        // less than the next expected external-eg-id)
-        if (live_updates->oldestEventGroupId() < last_global_eg_id_read + 1 || (newest_eg_id < end + 1)) {
+      if (not is_update_available) continue;
+
+      // Drop all live updates with global_eg_id < next_global_eg_id_to_read, because TRS has already read and sent
+      // these updates from storage
+      if (live_updates->oldestEventGroupId() < next_global_eg_id_to_read) {
+        SubEventGroupUpdate sub_eg_update;
+        live_updates->PopEventGroup(sub_eg_update);
+        LOG_INFO(logger_, "Sync dropping " << sub_eg_update.event_group_id);
+        is_update_available = false;
+        continue;
+      }
+
+      auto ext_eg_id = end + 1;  // next external event group ID to be read
+
+      // live_updates in sync with KVB i.e., event group ID associated with the oldest live update is the
+      // next_global_eg_id_to_read.
+      if (live_updates->oldestEventGroupId() == next_global_eg_id_to_read) {
+        auto filtered_eg_update = kvb_filter->filterEventGroupUpdate(live_updates->oldestEventGroup());
```

Comment:
We will read and filter the event group twice if this an event group for the client. Once here, and once after we return from this function.
Let's add a comment or TODO so we can take a look at potentially optimizing this later.